networkx/utils/backends.py

```diff
@@ -0,0 +1,628 @@
+        raise KeyError("'graphs' must contain at least one variable name") from None
+
+    # This dict comprehension is complicated for better performance; equivalent shown below.
+    optional_graphs = set()
+    list_graphs = set()
+    graphs = {
+        optional_graphs.add(val := k[:-1]) or val
+        if (last := k[-1]) == "?"
+        else list_graphs.add(val := k[1:-1]) or val
+        if last == "]"
+        else k: v
+        for k, v in graphs.items()
+    }
+    # The above is equivalent to:
+    # optional_graphs = {k[:-1] for k in graphs if k[-1] == "?"}
+    # list_graphs = {k[1:-1] for k in graphs if k[-1] == "]"}
+    # graphs = {k[:-1] if k[-1] == "?" else k: v for k, v in graphs.items()}
+
+    sig = inspect.signature(func)
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwds):
+        backend = plugins[plugin_name].load()
+        if not hasattr(backend, name):
+            if plugin_name == "nx-loopback":
+                raise NetworkXNotImplemented(
+                    f"'{name}' not found in {backend.__class__.__name__}"
+                )
+            pytest.xfail(f"'{name}' not implemented by {plugin_name}")
+        bound = sig.bind(*args, **kwds)
+        bound.apply_defaults()
+        # Check that graph names are actually in the signature
+        if set(graphs) - set(bound.arguments):
+            raise KeyError(f"Invalid graph names: {set(graphs) - set(bound.arguments)}")
+        # Convert graphs into backend graph-like object
+        #   Include the edge and/or node labels if provided to the algorithm
+        _preserve_edge_attrs = preserve_edge_attrs
+        _edge_attrs = edge_attrs
+        if _preserve_edge_attrs is False:
+            pass
+        elif _preserve_edge_attrs is True:
+            _edge_attrs = None
+        elif isinstance(_preserve_edge_attrs, str):
+            if bound.arguments[_preserve_edge_attrs] is True or callable(
+                bound.arguments[_preserve_edge_attrs]
+            ):
+                _preserve_edge_attrs = True
+                _edge_attrs = None
+            else:
+                _preserve_edge_attrs = False
+        elif not isinstance(_preserve_edge_attrs, dict):
+            raise TypeError(
+                f"Bad type for preserve_edge_attrs: {type(preserve_edge_attrs)}"
+            )
+
+        if _edge_attrs is None:
+            pass
+        elif isinstance(_edge_attrs, str):
+            if _edge_attrs[0] == "[":
+                _edge_attrs = {
+                    edge_attr: 1 for edge_attr in bound.arguments[_edge_attrs[1:-1]]
+                }
+            elif callable(bound.arguments[_edge_attrs]):
+                _preserve_edge_attrs = True
+                _edge_attrs = None
+            elif bound.arguments[_edge_attrs] is not None:
+                _edge_attrs = {bound.arguments[_edge_attrs]: 1}
+            elif name == "to_numpy_array" and hasattr(
+                bound.arguments["dtype"], "names"
+            ):
```

Comment:
So far, `to_numpy_array` here is the only "special case" that we handle in this function.